# Setup and Load Data

In [1]:
import numpy as np
import pandas as pd

In [2]:
def read_n_create_index(cityname):
    data = pd.read_csv("rent_listings_"+cityname+".csv")
    data['city_id'] = cityname + '_' +  data.index.astype(str)
    data.set_index('city_id', inplace=True)
    print("Size of the data:", data.shape)
    display(data.head())
    return data

In [3]:
lagos_data = read_n_create_index('lagos')

Size of the data: (5000, 6)


neighborhood        price  \
city_id                                                                   
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  ₦ 2,500,000   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  ₦ 2,500,000   
lagos_2                             -Old Ikoyi Ikoyi Lagos     $ 40,000   
lagos_3                       ...Lekki Phase 1 Lekki Lagos    ₦ 800,000   
lagos_4                          Lekki Phase 1 Lekki Lagos  ₦ 1,200,000   

           beds    baths    toilets  \
city_id                               
lagos_0  3 beds  4 baths  4 Toilets   
lagos_1  3 beds  3 baths  4 Toilets   
lagos_2  3 beds  3 baths  4 Toilets   
lagos_3  1 beds  1 baths  1 Toilets   
lagos_4  1 beds  1 baths  1 Toilets   

                                                     extra  
city_id                                                     
lagos_0  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_2  Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_3  For more information, Pricing, and Private Vie...  
lagos_4  For more information, Pricing, and Private Vie...

In [4]:
abuja_data = read_n_create_index('abuja')

Size of the data: (739, 6)


neighborhood        price    beds  \
city_id                                                                         
abuja_0       Sam najuma estate GaladimawaLokogoma Abuja  ₦ 1,300,000  3 beds   
abuja_1                             DawakiGwarinpa Abuja  ₦ 1,500,000  3 beds   
abuja_2                                     Gaduwa Abuja  ₦ 1,500,000  3 beds   
abuja_3                                    Maitama Abuja  ₦ 2,500,000  2 beds   
abuja_4  In an estate behind games village Garki 1 Abuja  ₦ 1,500,000  2 beds   

           baths    toilets                                              extra  
city_id                                                                         
abuja_0  3 baths  4 Toilets  Sharp 2nd floor 3 bedroom apartment @ Dr Sam n...  
abuja_1  3 baths  4 Toilets  Newly BuiltIt's newly built and never been occ...  
abuja_2  3 baths  4 Toilets  3bedroom blocks of flat in Gaduwa District, gr...  
abuja_3  2 baths  4 Toilets  ServicedA Serviced 2 bedroom flat in #maitama ...  
abuja_4  2 baths  3 Toilets  2 Bedroom Flat With CCTV, Constant Water Suppl...

In [5]:
ogun_data = read_n_create_index('ogun')

Size of the data: (188, 6)


neighborhood      price    beds    baths  \
city_id                                                                         
ogun_0             ShimawaMowe Obafemi Owode Ogun  ₦ 260,000  3 beds  1 baths   
ogun_1                       Adigbe Abeokuta Ogun  ₦ 300,000  2 beds  3 baths   
ogun_2   Private EstateMagboro Obafemi Owode Ogun  ₦ 250,000  1 beds  1 baths   
ogun_3          MagboroMagboro Obafemi Owode Ogun  ₦ 350,000  2 beds  2 baths   
ogun_4                   Private EstateArepo Ogun  ₦ 450,000  2 beds  2 baths   

           toilets                                              extra  
city_id                                                                
ogun_0   2 Toilets  Newly BuiltA newly built 1 bed, 2 beds and 3 b...  
ogun_1   3 Toilets  2 bedroom Shared Apartment Flat / Apartment fo...  
ogun_2   2 Toilets  Spacious Rooms Ample Parking Space Kitchen Cab...  
ogun_3   2 Toilets  24-hours Security Spacious Rooms Ample Parking...  
ogun_4   3 Toilets  24-hours Security Spacious Rooms Ample Parking...

In [6]:
ibadan_data = read_n_create_index('oyo')

Size of the data: (512, 6)


neighborhood        price  \
city_id                                                                   
oyo_0     Iyaganku GRA quarters IbadanIyanganku Ibadan Oyo  ₦ 1,000,000   
oyo_1    Behind DSS ile tuntun nihort Jericho ibadanIdi...    ₦ 450,000   
oyo_2     Yidi apete junction ibadan Ibadan Polytechnic...    ₦ 126,000   
oyo_3    Adeoyo GRA, off ring road, ibadan Ring Rd Ibad...    ₦ 750,000   
oyo_4    Opposition nnpc filling station, akala express...    ₦ 600,000   

           beds    baths    toilets  \
city_id                               
oyo_0    3 beds  3 baths  3 Toilets   
oyo_1    3 beds  3 baths  3 Toilets   
oyo_2    1 beds  1 baths  1 Toilets   
oyo_3    3 beds  3 baths  4 Toilets   
oyo_4    3 beds  3 baths  4 Toilets   

                                                     extra  
city_id                                                     
oyo_0    *To let Three bedroom flat en suit is availabl...  
oyo_1    *To let* 4 flat of 3 bedroom with 3 toilets, b...  
oyo_2    Serviced Newly BuiltA room selfcon at Yidi ape...  
oyo_3    Newly BuiltFinely finished 3 bedroom Apartment...  
oyo_4    Newly BuiltBrand new 3 bedroom Apartment for r...

In [7]:
ph_data = read_n_create_index('rivers')

Size of the data: (105, 6)


neighborhood        price  \
city_id                                                                    
rivers_0  GRA phase 3 tombia extensionNew Gra Port Harco...  ₦ 1,500,000   
rivers_1                    Ada George Port Harcourt Rivers    ₦ 500,000   
rivers_2         odili roadTrans Amadi Port Harcourt Rivers    ₦ 550,000   
rivers_3                   Trans Amadi Port Harcourt Rivers    ₦ 650,000   
rivers_4                ...Trans Amadi Port Harcourt Rivers    ₦ 800,000   

            beds    baths    toilets  \
city_id                                
rivers_0  2 beds  2 baths  2 Toilets   
rivers_1  2 beds  2 baths  2 Toilets   
rivers_2  1 beds  1 baths  2 Toilets   
rivers_3    beds    baths    Toilets   
rivers_4  0 beds  0 baths  0 Toilets   

                                                      extra  
city_id                                                      
rivers_0  Newly Built2no newly built 2bed room flats in ...  
rivers_1  Furnished Newly BuiltA VIRGIN TWO BEDROOM FLAT...  
rivers_2  One bedroom for [email protected] #550k odili ...  
rivers_3  Decent King Size 2 Bedroom Flat For Rent In Wo...  
rivers_4  King Size 2 Bedroom For Rent In Woji  Port Har...

# Data Lookup and Cleaning

In [8]:
data = pd.concat([lagos_data, abuja_data, ibadan_data, ogun_data, ph_data])
print("Size of the data:", data.shape)

data.head()

Size of the data: (6544, 6)


neighborhood        price  \
city_id                                                                   
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  ₦ 2,500,000   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  ₦ 2,500,000   
lagos_2                             -Old Ikoyi Ikoyi Lagos     $ 40,000   
lagos_3                       ...Lekki Phase 1 Lekki Lagos    ₦ 800,000   
lagos_4                          Lekki Phase 1 Lekki Lagos  ₦ 1,200,000   

           beds    baths    toilets  \
city_id                               
lagos_0  3 beds  4 baths  4 Toilets   
lagos_1  3 beds  3 baths  4 Toilets   
lagos_2  3 beds  3 baths  4 Toilets   
lagos_3  1 beds  1 baths  1 Toilets   
lagos_4  1 beds  1 baths  1 Toilets   

                                                     extra  
city_id                                                     
lagos_0  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_2  Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_3  For more information, Pricing, and Private Vie...  
lagos_4  For more information, Pricing, and Private Vie...

## Cleaning Price;

In [9]:
data.price.unique()

array(['₦ 2,500,000', '$ 40,000', '₦ 800,000', '₦ 1,200,000',
       '₦ 1,000,000', '₦ 350,000', '₦ 550,000', '₦ 700,000',
       '₦ 2,000,000', '₦ 500,000', '₦ 400,000', '₦ 600,000', '₦ 300,000',
       '₦ 900,000', '₦ 1,500,000', '₦ 1,800,000', '₦ 850,000',
       '₦ 170,000', '₦ 1,600,000', '₦ 650,000', '₦ 750,000',
       '₦ 1,300,000', '₦ 1,100,000', '₦ 2,300,000', '₦ 250,000',
       '₦ 1,400,000', '₦ 450,000', '₦ 200,000', '₦ 120,000',
       '₦ 1,700,000', '₦ 950,000', '₦ 55,000', '₦ 130,000', '₦ 1,900,000',
       '₦ 65,000', '₦ 1,850,000', '₦ 150,000', '₦ 1,650,000', '₦ 470,000',
       '₦ 1,350,000', '₦ 270,000', '₦ 35,000', '$ 25,000', '$ 65,000',
       '₦ 2,200,000', '$ 100,000', '₦ 2,250,000', '$ 80,000', '$ 28,000',
       '₦ 70,000', '$ 50,000', '₦ 2,100,000', '₦ 840,000', '$ 70,000',
       '₦ 210,000', '₦ 216,000', '₦ 250', '₦ 84,000', '₦ 160,000',
       '₦ 1,025,000', '₦ 230,000', '₦ 280,000', '₦ 480,000', '₦ 430,000',
       '₦ 220,000', '₦ 2,400,000', '₦ 330,000'

Some of the prices are in dollars! We'd probably drop those rows.

In [10]:
dollar_apartments = data[data.price.str.startswith('$')]
dollar_apartments

neighborhood      price  \
city_id                                                                    
lagos_2                                -Old Ikoyi Ikoyi Lagos   $ 40,000   
lagos_398                               Victoria Island Lagos   $ 25,000   
lagos_399                               Victoria Island Lagos   $ 65,000   
lagos_404                               Old Ikoyi Ikoyi Lagos  $ 100,000   
lagos_408                               Old Ikoyi Ikoyi Lagos   $ 80,000   
lagos_412                               Victoria Island Lagos   $ 28,000   
lagos_433                               Old Ikoyi Ikoyi Lagos   $ 50,000   
lagos_605                                         Ikoyi Lagos   $ 70,000   
lagos_1689     Ozumba MbadiweBonny Camp Victoria Island Lagos   $ 75,000   
lagos_1874                -Eko Atlantic Victoria Island Lagos   $ 50,000   
lagos_2268              Ligali Ayorinde Victoria Island Lagos   $ 80,000   
lagos_2294  victoria island Victoria Island Victoria Islan...   $ 65,000   
lagos_2351  YOUNIS BASHORUNVictoria Island Extension Victo...   $ 80,000   
lagos_2504            Banana Island Banana Island Ikoyi Lagos   $ 90,000   
lagos_2572    Adeola Odeku Kofo Abayomi Victoria Island Lagos   $ 25,000   
lagos_2581                 Eko Atlantic Victoria Island Lagos   $ 90,000   
lagos_2800             Banana IslandBanana Island Ikoyi Lagos   $ 90,000   
lagos_3130  Kofo Abayomi Ahmadu Bello Way Victoria Island ...   $ 12,000   
lagos_4082                 Eko Atlantic Victoria Island Lagos   $ 70,000   
lagos_4893             Banana IslandBanana Island Ikoyi Lagos   $ 65,000   
lagos_4979               VIAkin Adesola Victoria Island Lagos   $ 80,000   
abuja_75                        Oyi river streetMaitama Abuja   $ 35,000   
abuja_81                        Oyi river StreetMaitama Abuja   $ 35,000   
abuja_124                          Osun crescentMaitama Abuja   $ 18,000   
abuja_140                               Maitama Maitama Abuja   $ 68,000   
abuja_143                                MaitamaMaitama Abuja   $ 25,000   
abuja_292                          Sunrise HillsAsokoro Abuja   $ 65,000   
abuja_305                          Sunrise HillsAsokoro Abuja   $ 65,000   
abuja_334                         Maitama, AbujaMaitama Abuja  $ 105,000   
abuja_546                               Asokoro Asokoro Abuja   $ 45,000   

              beds    baths     toilets  \
city_id                                   
lagos_2     3 beds  3 baths   4 Toilets   
lagos_398   3 beds  3 baths   4 Toilets   
lagos_399   3 beds  3 baths   4 Toilets   
lagos_404   3 beds  3 baths   4 Toilets   
lagos_408   3 beds  3 baths   4 Toilets   
lagos_412   1 beds  1 baths   2 Toilets   
lagos_433   3 beds  3 baths   4 Toilets   
lagos_605   3 beds  3 baths   4 Toilets   
lagos_1689  3 beds  3 baths   4 Toilets   
lagos_1874  3 beds  3 baths   4 Toilets   
lagos_2268  3 beds  4 baths   4 Toilets   
lagos_2294  3 beds  0 baths   0 Toilets   
lagos_2351  3 beds  3 baths   4 Toilets   
lagos_2504  3 beds  3 baths   4 Toilets   
lagos_2572  1 beds  2 baths   2 Toilets   
lagos_2581  3 beds  3 baths   4 Toilets   
lagos_2800  3 beds  3 baths   4 Toilets   
lagos_3130  1 beds  1 baths   1 Toilets   
lagos_4082  4 beds  4 baths   5 Toilets   
lagos_4893  4 beds  4 baths   5 Toilets   
lagos_4979  3 beds  3 baths   4 Toilets   
abuja_75    2 beds  2 baths   3 Toilets   
abuja_81    2 beds  2 baths   3 Toilets   
abuja_124   3 beds  3 baths   4 Toilets   
abuja_140   3 beds  3 baths   4 Toilets   
abuja_143   3 beds  3 baths   4 Toilets   
abuja_292   3 beds  3 baths   4 Toilets   
abuja_305   3 beds  3 baths   4 Toilets   
abuja_334   4 beds  4 baths   5 Toilets   
abuja_546   3 beds  3 baths   4 Toilets   

                                                        extra  
city_id                                                        
lagos_2     Serviced Newly Built3 bedroom Flat / Apartment...  
lagos_398   Modern spacious flat in Victoria Is

They all seems to be located at either Lagos or Abuja.

In [11]:
data.drop(dollar_apartments.index, inplace=True)

data.price.unique()

array(['₦ 2,500,000', '₦ 800,000', '₦ 1,200,000', '₦ 1,000,000',
       '₦ 350,000', '₦ 550,000', '₦ 700,000', '₦ 2,000,000', '₦ 500,000',
       '₦ 400,000', '₦ 600,000', '₦ 300,000', '₦ 900,000', '₦ 1,500,000',
       '₦ 1,800,000', '₦ 850,000', '₦ 170,000', '₦ 1,600,000',
       '₦ 650,000', '₦ 750,000', '₦ 1,300,000', '₦ 1,100,000',
       '₦ 2,300,000', '₦ 250,000', '₦ 1,400,000', '₦ 450,000',
       '₦ 200,000', '₦ 120,000', '₦ 1,700,000', '₦ 950,000', '₦ 55,000',
       '₦ 130,000', '₦ 1,900,000', '₦ 65,000', '₦ 1,850,000', '₦ 150,000',
       '₦ 1,650,000', '₦ 470,000', '₦ 1,350,000', '₦ 270,000', '₦ 35,000',
       '₦ 2,200,000', '₦ 2,250,000', '₦ 70,000', '₦ 2,100,000',
       '₦ 840,000', '₦ 210,000', '₦ 216,000', '₦ 250', '₦ 84,000',
       '₦ 160,000', '₦ 1,025,000', '₦ 230,000', '₦ 280,000', '₦ 480,000',
       '₦ 430,000', '₦ 220,000', '₦ 2,400,000', '₦ 330,000', '₦ 180,000',
       '₦ 530,000', '₦ 10,000', '₦ 380,000', '₦ 420,000', '₦ 90,000',
       '₦ 110,000', '₦ 500

No dollar-priced apartments anymore!

In [12]:
data.price = data.price.str.replace(',', '').apply(lambda x: x[2:])

data.price.unique()

array(['2500000', '800000', '1200000', '1000000', '350000', '550000',
       '700000', '2000000', '500000', '400000', '600000', '300000',
       '900000', '1500000', '1800000', '850000', '170000', '1600000',
       '650000', '750000', '1300000', '1100000', '2300000', '250000',
       '1400000', '450000', '200000', '120000', '1700000', '950000',
       '55000', '130000', '1900000', '65000', '1850000', '150000',
       '1650000', '470000', '1350000', '270000', '35000', '2200000',
       '2250000', '70000', '2100000', '840000', '210000', '216000', '250',
       '84000', '160000', '1025000', '230000', '280000', '480000',
       '430000', '220000', '2400000', '330000', '180000', '530000',
       '10000', '380000', '420000', '90000', '110000', '500', '1150000',
       '140000', '1250000', '25000', '100000', '96000', '40000', '45000',
       '460000', '630000', '72000', '370000', '6', '620000', '1200009',
       '50000', '320000', '1750000', '60000', '270', '150', '350',
       '190000', '520

There are instances where the price value is less than 4 digits. Having a closer look;

In [13]:
cheap_apartments = data[data.price.str.len() <= 5]
cheap_apartments

neighborhood  price    beds  \
city_id                                                                        
lagos_178                           Lekki Phase 1 Lekki Lagos  55000  3 beds   
lagos_204                             ...Abule Oja Yaba Lagos  65000  1 beds   
lagos_397   Shonibare Estate Mobolaji Bank Anthony Way Ike...  35000  1 beds   
lagos_430                         Parkview Estate Ikoyi Lagos  70000  3 beds   
lagos_802   Madonna Estate ojodu Abiodun road Berger Ojodu...    250  1 beds   
lagos_820    Unity street,unity bus stopIgbogbo Ikorodu Lagos  84000  1 beds   
lagos_1344                                   Akoka Yaba Lagos  10000    beds   
lagos_1696  Itele ogun state close to ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_1753                           Ojodu Berger Ojodu Lagos    500  3 beds   
lagos_1866                              Phase 2 Gbagada Lagos  25000  1 beds   
lagos_2071  Nasfat street before Agunfoye sawmill along ig...  96000  1 beds   
lagos_2136  Off freedom road lekki phase 1Lekki Phase 1 Le...  40000  1 beds   
lagos_2213  Off studio 24, Lekki phase 1 Lekki Phase 1 Lek...  45000  2 beds   
lagos_2346  Itele Ogun state close to Ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_2363  Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
lagos_2493             Rumens Road IkoyiOld Ikoyi Ikoyi Lagos      6  1 beds   
lagos_2502                              Igbogbo Ikorodu Lagos  90000  1 beds   
lagos_2543                Off Lekki Epe Expressway Ajah Lagos  70000  2 beds   
lagos_2615  Very decent and beautiful new house 2bedroom a...  50000  2 beds   
lagos_2653  Itele Ogun State close to Ayobo LagosAyobo Ipa...  90000  1 beds   
lagos_2697                         Spar roadIkate Lekki Lagos  45000  1 beds   
lagos_2839  Asekun street off adekunle kuyeAdelabu Suruler...    500  1 beds   
lagos_3112  Off Admirathy road Lekki phase 1 Lekki Phase 1...  25000  1 beds   
lagos_3143                  Itele ayobo roadAyobo Ipaja Lagos  60000  1 beds   
lagos_3194  Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
lagos_3208  Okotie Eboh street off Awolowo road ikoyiAwolo...  40000  1 beds   
lagos_3222                 College road Ifako Ogba Ogba Lagos    270    beds   
lagos_3225        3 Greenville estate badoreBadore Ajah Lagos    150  1 beds   
lagos_3241          Abule-ijesha road Abule Ijesha Yaba Lagos    350    beds   
lagos_3320                          Lekki Phase 1 Lekki Lagos  50000  3 beds   
lagos_4221              Commissioner streetIlasan Lekki Lagos    600  1 beds   
lagos_4334           Ifako GbagadaIfako Gbagada Gbagada Lagos      1  3 beds   
lagos_4510                    TransformerOkun Ajah Ajah Lagos   4000  1 beds   
abuja_295                              GwarinpaGwarinpa Abuja   1300  1 beds   
abuja_495          BY TIMBER SHADE, BWARI, Central Area Abuja  90000  1 beds   
abuja_631                                      ...Lugbe Abuja    600  1 beds   
abuja_688               Wuse Zone 7 Harper StreetWuse 2 Abuja  90000  6 beds   
oyo_304                                            Ibadan Oyo  80000  1 beds   
oyo_449     U.i-Ojoo old bank road expressway IbadanIbadan...  80000  1 beds   
oyo_462      Alawaye street off akanran roadIwo Rd Ibadan Oyo  75009  1 beds   
oyo_501     Alakia old ife road inside airport second gate...  60000  1 beds   
ogun_18     Itele ogun state close to ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_31     Itele Ogun state close to Ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_32     Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
ogun_36     Itele Ogun State close to Ayobo LagosAyobo Ipa...  90000  1 beds   
ogun_47     Itele Ogun State close to Ayobo LagosAyobo Ipa...  72000  1 beds   
rivers_94   Golf Estate, end of Peter Odili RoadTrans Amad...  35000  2 beds   

              baths    toilets  \
city_id                          
lagos_178   4 baths  4 Toilets   
lagos_204   1 baths  1 Toi

Whilst the apartments in Ogun and Ibadan might be truly cheap, those in Lagos and Abuja are almost certainly due to data inconsistencies.

In [14]:
exempted_towns = ['korodu', 'ogun', 'Ayobo', 'Ibadan']
for town in exempted_towns:
    cheap_apartments = cheap_apartments[~cheap_apartments.neighborhood.str.contains(town)]
cheap_apartments

neighborhood  price    beds  \
city_id                                                                        
lagos_178                           Lekki Phase 1 Lekki Lagos  55000  3 beds   
lagos_204                             ...Abule Oja Yaba Lagos  65000  1 beds   
lagos_397   Shonibare Estate Mobolaji Bank Anthony Way Ike...  35000  1 beds   
lagos_430                         Parkview Estate Ikoyi Lagos  70000  3 beds   
lagos_802   Madonna Estate ojodu Abiodun road Berger Ojodu...    250  1 beds   
lagos_1344                                   Akoka Yaba Lagos  10000    beds   
lagos_1753                           Ojodu Berger Ojodu Lagos    500  3 beds   
lagos_1866                              Phase 2 Gbagada Lagos  25000  1 beds   
lagos_2136  Off freedom road lekki phase 1Lekki Phase 1 Le...  40000  1 beds   
lagos_2213  Off studio 24, Lekki phase 1 Lekki Phase 1 Lek...  45000  2 beds   
lagos_2493             Rumens Road IkoyiOld Ikoyi Ikoyi Lagos      6  1 beds   
lagos_2543                Off Lekki Epe Expressway Ajah Lagos  70000  2 beds   
lagos_2615  Very decent and beautiful new house 2bedroom a...  50000  2 beds   
lagos_2697                         Spar roadIkate Lekki Lagos  45000  1 beds   
lagos_2839  Asekun street off adekunle kuyeAdelabu Suruler...    500  1 beds   
lagos_3112  Off Admirathy road Lekki phase 1 Lekki Phase 1...  25000  1 beds   
lagos_3208  Okotie Eboh street off Awolowo road ikoyiAwolo...  40000  1 beds   
lagos_3222                 College road Ifako Ogba Ogba Lagos    270    beds   
lagos_3225        3 Greenville estate badoreBadore Ajah Lagos    150  1 beds   
lagos_3241          Abule-ijesha road Abule Ijesha Yaba Lagos    350    beds   
lagos_3320                          Lekki Phase 1 Lekki Lagos  50000  3 beds   
lagos_4221              Commissioner streetIlasan Lekki Lagos    600  1 beds   
lagos_4334           Ifako GbagadaIfako Gbagada Gbagada Lagos      1  3 beds   
lagos_4510                    TransformerOkun Ajah Ajah Lagos   4000  1 beds   
abuja_295                              GwarinpaGwarinpa Abuja   1300  1 beds   
abuja_495          BY TIMBER SHADE, BWARI, Central Area Abuja  90000  1 beds   
abuja_631                                      ...Lugbe Abuja    600  1 beds   
abuja_688               Wuse Zone 7 Harper StreetWuse 2 Abuja  90000  6 beds   
rivers_94   Golf Estate, end of Peter Odili RoadTrans Amad...  35000  2 beds   

              baths    toilets  \
city_id                          
lagos_178   4 baths  4 Toilets   
lagos_204   1 baths  1 Toilets   
lagos_397   1 baths  2 Toilets   
lagos_430   3 baths  4 Toilets   
lagos_802   1 baths  1 Toilets   
lagos_1344    baths    Toilets   
lagos_1753  3 baths  2 Toilets   
lagos_1866  1 baths  1 Toilets   
lagos_2136  1 baths  2 Toilets   
lagos_2213  1 baths  2 Toilets   
lagos_2493  2 baths  2 Toilets   
lagos_2543  2 baths  3 Toilets   
lagos_2615  2 baths  2 Toilets   
lagos_2697  1 baths  2 Toilets   
lagos_2839  1 baths  1 Toilets   
lagos_3112  1 baths  1 Toilets   
lagos_3208  1 baths  2 Toilets   
lagos_3222    baths    Toilets   
lagos_3225  1 baths  1 Toilets   
lagos_3241    baths    Toilets   
lagos_3320  3 baths  4 Toilets   
lagos_4221  1 baths  1 Toilets   
lagos_4334  3 baths  3 Toilets   
lagos_4510  1 baths  1 Toilets   
abuja_295   1 baths  2 Toilets   
abuja_495   0 baths  0 Toilets   
abuja_631   2 baths  2 Toilets   
abuja_688   2 baths  2 Toilets   
rivers_94   2 baths  2 Toilets   

                                                        extra  
city_id                                                        
lagos_178   Furnished Serviced Newly BuiltThis is a shorle...  
lagos_204   ServicedSERVICED & SPACIOUS ROOMSELF FOR RENT....  
lagos_397   Furnished ServicedThe property is one bedroom ...  
lagos_430   Furnished3 bedroom Flat / Apartment for rent P...  
lagos_802   FurnishedA room self contain for rent at madon...  
lagos_1344  HOT SHORT-LET ?????? serviced Studio Apartment...  
l

Now we can drop the truly ridiculous ones;

In [15]:
data.drop(cheap_apartments.index, inplace=True)
data.head()

neighborhood    price    beds  \
city_id                                                                       
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  2500000  3 beds   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  2500000  3 beds   
lagos_3                       ...Lekki Phase 1 Lekki Lagos   800000  1 beds   
lagos_4                          Lekki Phase 1 Lekki Lagos  1200000  1 beds   
lagos_5                                         Ketu Lagos  1000000  2 beds   

           baths    toilets                                              extra  
city_id                                                                         
lagos_0  4 baths  4 Toilets  Brand new service 3 bedroom flat all rooms ens...  
lagos_1  3 baths  4 Toilets  The property is a 3 Bedrooms Upscale FLAT all ...  
lagos_3  1 baths  1 Toilets  For more information, Pricing, and Private Vie...  
lagos_4  1 baths  1 Toilets  For more information, Pricing, and Private Vie...  
lagos_5  2 baths  2 Toilets  Brand new 2 bedrooms flat for rent at Ketu Lag...

## Cleaning Beds;

In [16]:
data.beds.unique()

array(['3 beds', '1 beds', '2 beds', 'beds', '4 beds', '0 beds', '5 beds',
       '6 beds', '3 beds ', '10 beds', 'beds ', '8 beds'], dtype=object)

Some beds don't have a digit value, while some have 0.

In [17]:
no_beds = data[data.beds.str.startswith('beds')]
no_beds

neighborhood    price  beds   baths    toilets  \
city_id                                                                         
lagos_7        Salem LekkiIkate Lekki Lagos  1000000  beds   baths    Toilets   
lagos_12           Akinola Akoka Yaba Lagos   400000  beds   baths  1 Toilets   
lagos_17        St.Finbars Akoka Yaba Lagos   500000  beds   baths  1 Toilets   
lagos_18       St. Finbars Akoka Yaba Lagos   600000  beds   baths  1 Toilets   
lagos_31            Lawanson Surulere Lagos   300000  beds   baths    Toilets   
...                                     ...      ...   ...     ...        ...   
rivers_84  Trans Amadi Port Harcourt Rivers   550000  beds   baths    Toilets   
rivers_85  Trans Amadi Port Harcourt Rivers   600000  beds   baths    Toilets   
rivers_86  Trans Amadi Port Harcourt Rivers   800000  beds   baths    Toilets   
rivers_88       Eliozu Port Harcourt Rivers   800000  beds   baths    Toilets   
rivers_90  Trans Amadi Port Harcourt Rivers   400000  beds   baths    Toilets   

                                                       extra  
city_id                                                       
lagos_7    Nice and Standard Renovated Mini Flat availabl...  
lagos_12   Furnished ServicedThis is a lovely SelfContain...  
lagos_17   Furnished Serviced Newly BuiltThis is a lovely...  
lagos_18   Furnished Serviced Newly BuiltThis is a lovely...  
lagos_31   *Clean Miniflat with tiles and water at idiara...  
...                                                      ...  
rivers_84  Cool 2 Bedroom Flat For Rent in Woji Port Harc...  
rivers_85  Modern King Size 2 Bedroom Flat For Rent In Wo...  
rivers_86  Tastefully Finished 2 Bedroom Flat For Rent in...  
rivers_88  Well Finished Brand new 2 Bedroom Flat For Ren...  
rivers_90  Decent Size 2 Bedroom Flat For Rent In Woji Po...  

[349 rows x 6 columns]

In [18]:
data.drop(no_beds.index, inplace=True)

data.beds.unique()

array(['3 beds', '1 beds', '2 beds', '4 beds', '0 beds', '5 beds',
       '6 beds', '3 beds ', '10 beds', '8 beds'], dtype=object)

In [19]:
data.beds = data.beds.map(lambda x: x[0]).astype(int)

data.beds.unique()

array([3, 1, 2, 4, 0, 5, 6, 8])

Great! Now onto the 0 beds;

In [20]:
data[data.beds == 0]

neighborhood    price  beds  \
city_id                                                                       
lagos_65   Close to 3rd Mainland Bridge, Yaba, Lagos. Yab...   800000     0   
lagos_212                       ...Adeniyi Jones Ikeja Lagos  2000000     0   
lagos_254                           Chemist Akoka Yaba Lagos   400000     0   
lagos_268         Obadia- off ilaje roadBariga Shomolu Lagos   350000     0   
lagos_441     destiny home estate Abijo AJAHAbijo Ajah Lagos   500000     0   
...                                                      ...      ...   ...   
rivers_61                ...Trans Amadi Port Harcourt Rivers   800000     0   
rivers_75                 - Trans Amadi Port Harcourt Rivers   750000     0   
rivers_82                ...Trans Amadi Port Harcourt Rivers   500000     0   
rivers_83                 - Trans Amadi Port Harcourt Rivers  1000000     0   
rivers_89                  -Trans Amadi Port Harcourt Rivers  1200000     0   

             baths    toilets  \
city_id                         
lagos_65   0 baths  0 Toilets   
lagos_212  2 baths  3 Toilets   
lagos_254  0 baths  1 Toilets   
lagos_268  0 baths  0 Toilets   
lagos_441  0 baths  0 Toilets   
...            ...        ...   
rivers_61  0 baths  0 Toilets   
rivers_75  0 baths  0 Toilets   
rivers_82  0 baths  0 Toilets   
rivers_83  0 baths  0 Toilets   
rivers_89  0 baths  0 Toilets   

                                                       extra  
city_id                                                       
lagos_65   Currently Being Renovated: Spacious Miniflat (...  
lagos_212  JUST OUT IN ADENIYI JONES 2 bedroom Luxery apa...  
lagos_254  ServicedThis is a lovely SelfContain @ Akoka Y...  
lagos_268  FurnishedA Newly Renovated Mini Flat @ Obadiah...  
lagos_441  Room and parlour self contain for rent at dest...  
...                                                      ...  
rivers_61  Modern King Size 3 Bedroom Flat For Rent In Wo...  
rivers_75  Beautiful 1 Bedroom Flat For Rent In Woji Port...  
rivers_82  Neat 2 Bedroom Flat For Rent in Woji Port Harc...  
rivers_83  Decent king size 2 Bedroom Flat For Rent In Pe...  
rivers_89  Decent 2 Bedroom Flat For Rent in Peter Odili ...  

[161 rows x 6 columns]

Some of them are just plain errors and are already in the description column. 

I'll convert every word in the `description` column to lowercase letters, for consistency.

In [21]:
data.extra = data.extra.str.lower()
data.head()

neighborhood    price  beds  \
city_id                                                                     
lagos_0         OLALEYE Estate Alaka Iponri Surulere Lagos  2500000     3   
lagos_1  Off Kudirat Abiola way, Julie EstateOregun Ike...  2500000     3   
lagos_3                       ...Lekki Phase 1 Lekki Lagos   800000     1   
lagos_4                          Lekki Phase 1 Lekki Lagos  1200000     1   
lagos_5                                         Ketu Lagos  1000000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_0  4 baths  4 Toilets  brand new service 3 bedroom flat all rooms ens...  
lagos_1  3 baths  4 Toilets  the property is a 3 bedrooms upscale flat all ...  
lagos_3  1 baths  1 Toilets  for more information, pricing, and private vie...  
lagos_4  1 baths  1 Toilets  for more information, pricing, and private vie...  
lagos_5  2 baths  2 Toilets  brand new 2 bedrooms flat for rent at ketu lag...

In [22]:
for i in range(1, 6):
    data.loc[data.extra.str.contains(str(i)), 'beds'] = i
data.loc[data.extra.str.contains('two'), 'beds'] = 2 # for those written in words
data.loc[data.extra.str.contains('three'), 'beds'] = 3 # for those written in words

data.loc[data.extra.str.contains('and parlour'), 'beds'] = 1 # for 'room and parlour' apartments
data.loc[data.extra.str.contains('self'), 'beds'] = 1 # for 'self contained' apartments
data.loc[data.extra.str.contains('mini'), 'beds'] = 1 # for mini flats

bed_zero = data[data.beds == 0]
bed_zero

neighborhood   price  beds  \
city_id                                                                       
lagos_500                            Harmony estateAjah Lagos  400000     0   
lagos_2118                  Bama villaAbule Ijesha Yaba Lagos  400000     0   
lagos_4053  ShopRite Road Osapa London LekkiOsapa London L...  350000     0   
lagos_4250        S Forte Estate, Before BlencoAdo Ajah Lagos  650000     0   
oyo_431          First Power line Eleyele/OloguneruIbadan Oyo  150000     0   
oyo_433                           Old BodijaBodija Ibadan Oyo  400000     0   

               baths     toilets  \
city_id                            
lagos_500    0 baths   0 Toilets   
lagos_2118   1 baths   1 Toilets   
lagos_4053   0 baths   0 Toilets   
lagos_4250   0 baths   0 Toilets   
oyo_431      0 baths   0 Toilets   
oyo_433      0 baths   0 Toilets   

                                                        extra  
city_id                                                        
lagos_500               very spacious and neat environment..   
lagos_2118  furnishedhot letting: lovely built serviced ap...  
lagos_4053  serviced newly builtnice and standard newly bu...  
lagos_4250  flat / apartment for rent s forte estate, befo...  
oyo_431     it is located in a well conducive and serene e...  
oyo_433     it is located in a well conducive and serene e...

And I've reduced 161 unknown rooms to just a handful!

In [23]:
data.drop(bed_zero.index, inplace=True)

data.beds.unique()

array([3, 1, 2, 5, 4])

## Cleaning Baths & Toilets;

Since most houses do have both bathrooms and toilets adjoined, except where there's an extra pair of toilet in the sitting room for visitors, I'd most likely drop `toilet` column and go with bathroom alone.

In [24]:
data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', ' baths', '0 baths',
       ' 2 baths', '5 baths', ' 3 baths', ' 1 baths', '  baths',
       '6 baths', '10 baths', ' 5 baths'], dtype=object)

In [25]:
data.baths = data.baths.str.strip() # cleaning the starting and trailing whitespaces

data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', 'baths', '0 baths',
       '5 baths', '6 baths', '10 baths'], dtype=object)

**Checking for toilets;**

In [26]:
data.toilets.unique()

array(['4 Toilets', '1 Toilets', '2 Toilets', '3 Toilets', ' Toilets',
       ' 4 Toilets', '0 Toilets', ' 1 Toilets', '5 Toilets', ' 0 Toilets',
       '6 Toilets', ' 3 Toilets', ' 2 Toilets', '10 Toilets', '  Toilets'],
      dtype=object)

In [27]:
data.toilets = data.toilets.str.strip() # cleaning the whitespaces

data.toilets.unique()

array(['4 Toilets', '1 Toilets', '2 Toilets', '3 Toilets', 'Toilets',
       '0 Toilets', '5 Toilets', '6 Toilets', '10 Toilets'], dtype=object)

**Checking for the ones without a digit;**

In [28]:
no_baths = data[data.baths.str.startswith('baths')]
no_baths

neighborhood    price  beds  \
city_id                                                                       
lagos_26                                         Agege Lagos   800000     3   
lagos_28            Off Demurin road Ketu Kosofe Ikosi Lagos   400000     1   
lagos_29            Off Demurin road Ketu Kosofe Ikosi Lagos  1000000     2   
lagos_30              Seaside estate badoreBadore Ajah Lagos  1800000     4   
lagos_68   Alagomeji-Adekunle, Close to the 3rd Mainland ...  1500000     2   
...                                                      ...      ...   ...   
ogun_108                                  Obafemi Owode Ogun   400000     2   
ogun_153                Arepo private EstateArepo Arepo Ogun   500000     2   
ogun_182   Agbeloba upright hotel street Abalawa Abeokuta...   250000     1   
rivers_81                   Trans Amadi Port Harcourt Rivers   400000     1   
rivers_87                        Eliozu Port Harcourt Rivers  1000000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_26   baths    Toilets  brand new 3 bedroom flat in oko oba, agege for...  
lagos_28   baths    Toilets  newly renovated upstairs roomself apartment pr...  
lagos_29   baths    Toilets  newly renovated 2 bedrooms flats with dining, ...  
lagos_30   baths    Toilets  4bedroom duplex, alone in the compound, pop fi...  
lagos_68   baths  2 Toilets  lovely new & very spacious executive 2 bedroom...  
...          ...        ...                                                ...  
ogun_108   baths    Toilets  furnisheda nice 2 bedroom apartment in a priva...  
ogun_153   baths  3 Toilets  furnished newly builtlovely 2 bed room flat at...  
ogun_182   baths  1 Toilets  newly builtroom and parlour self contained wit...  
rivers_81  baths    Toilets  tastefully finished new 1 bedroom flat for ren...  
rivers_87  baths    Toilets  decent brand new 2 bedroom flat for rent in sh...  

[317 rows x 6 columns]

In [29]:
data[data.toilets.str.startswith('T')]

neighborhood    price  beds    baths  \
city_id                                                                       
lagos_26                                Agege Lagos   800000     3    baths   
lagos_28   Off Demurin road Ketu Kosofe Ikosi Lagos   400000     1    baths   
lagos_29   Off Demurin road Ketu Kosofe Ikosi Lagos  1000000     2    baths   
lagos_30     Seaside estate badoreBadore Ajah Lagos  1800000     4    baths   
lagos_161                      Masha Surulere Lagos  1600000     3  2 baths   
...                                             ...      ...   ...      ...   
rivers_70          Trans Amadi Port Harcourt Rivers  1100000     3  3 baths   
rivers_72         -Trans Amadi Port Harcourt Rivers  1500000     3  3 baths   
rivers_76          Trans Amadi Port Harcourt Rivers  1000000     3  3 baths   
rivers_81          Trans Amadi Port Harcourt Rivers   400000     1    baths   
rivers_87               Eliozu Port Harcourt Rivers  1000000     2    baths   

           toilets                                              extra  
city_id                                                                
lagos_26   Toilets  brand new 3 bedroom flat in oko oba, agege for...  
lagos_28   Toilets  newly renovated upstairs roomself apartment pr...  
lagos_29   Toilets  newly renovated 2 bedrooms flats with dining, ...  
lagos_30   Toilets  4bedroom duplex, alone in the compound, pop fi...  
lagos_161  Toilets  servicedfor rent... a serviced & executive 3be...  
...            ...                                                ...  
rivers_70  Toilets  spacious 3 bedroom flat for rent in peter odil...  
rivers_72  Toilets  tastefully built 3 bedroom flat for rent in pe...  
rivers_76  Toilets  executive newly built 3 bedroom flat for rent ...  
rivers_81  Toilets  tastefully finished new 1 bedroom flat for ren...  
rivers_87  Toilets  decent brand new 2 bedroom flat for rent in sh...  

[392 rows x 6 columns]

I could fill replace the baths without a digit value with their corresponding toilet values, provided that their toilet has a digit value.

In [30]:
data.loc[data.baths.str.startswith('baths'), 'baths'] = data.toilets.map(lambda x: x.split(' ')[0])

data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', 'Toilets', '0 baths',
       '2', '1', '3', '5 baths', '4', '6 baths', '10 baths'], dtype=object)

Great! But the ones with no-digit toilet values got replaced by 'Toilet'...

In [31]:
no_digit_baths = data[data.baths.str.startswith('Toilets')]
no_digit_baths

neighborhood    price  beds  \
city_id                                                                       
lagos_26                                         Agege Lagos   800000     3   
lagos_28            Off Demurin road Ketu Kosofe Ikosi Lagos   400000     1   
lagos_29            Off Demurin road Ketu Kosofe Ikosi Lagos  1000000     2   
lagos_30              Seaside estate badoreBadore Ajah Lagos  1800000     4   
lagos_191  Akinola Cole, Adebiyi Jones Ikeja.Adeniyi Jone...  1600000     3   
...                                                      ...      ...   ...   
oyo_487                 Beeshop areaAkala Express Ibadan Oyo   800000     3   
ogun_100                          Magboro Obafemi Owode Ogun   600000     5   
ogun_108                                  Obafemi Owode Ogun   400000     2   
rivers_81                   Trans Amadi Port Harcourt Rivers   400000     1   
rivers_87                        Eliozu Port Harcourt Rivers  1000000     2   

             baths  toilets                                              extra  
city_id                                                                         
lagos_26   Toilets  Toilets  brand new 3 bedroom flat in oko oba, agege for...  
lagos_28   Toilets  Toilets  newly renovated upstairs roomself apartment pr...  
lagos_29   Toilets  Toilets  newly renovated 2 bedrooms flats with dining, ...  
lagos_30   Toilets  Toilets  4bedroom duplex, alone in the compound, pop fi...  
lagos_191  Toilets  Toilets  1 units service 3 bedroom apartment at akinola...  
...            ...      ...                                                ...  
oyo_487    Toilets  Toilets  2 bedrooms flat with pop, 3 toilets, 3 ac, wat...  
ogun_100   Toilets  Toilets  furnished newly builtnewly built 3 bed room fl...  
ogun_108   Toilets  Toilets  furnisheda nice 2 bedroom apartment in a priva...  
rivers_81  Toilets  Toilets  tastefully finished new 1 bedroom flat for ren...  
rivers_87  Toilets  Toilets  decent brand new 2 bedroom flat for rent in sh...  

[276 rows x 6 columns]

I managed to save about 40 rows of data. I'll drop the rest.

In [32]:
data.drop(no_digit_baths.index, inplace=True)

data.baths.unique()

array(['4 baths', '3 baths', '1 baths', '2 baths', '0 baths', '2', '1',
       '3', '5 baths', '4', '6 baths', '10 baths'], dtype=object)

In [33]:
data.baths = data.baths.apply(lambda x: x.split(' ')[0]).astype(int) # converting baths to integers

data.baths.unique()

array([ 4,  3,  1,  2,  0,  5,  6, 10])

Now treating the ones with 0 as digit;

In [34]:
data[data.baths == 0]

neighborhood    price  beds  \
city_id                                                                       
lagos_65   Close to 3rd Mainland Bridge, Yaba, Lagos. Yab...   800000     1   
lagos_254                           Chemist Akoka Yaba Lagos   400000     1   
lagos_268         Obadia- off ilaje roadBariga Shomolu Lagos   350000     1   
lagos_441     destiny home estate Abijo AJAHAbijo Ajah Lagos   500000     1   
lagos_442  Ajiran Road, Agungi, off Lekki-Epe Express Way...  2000000     3   
...                                                      ...      ...   ...   
rivers_71               ....Trans Amadi Port Harcourt Rivers  1000000     1   
rivers_75                 - Trans Amadi Port Harcourt Rivers   750000     1   
rivers_82                ...Trans Amadi Port Harcourt Rivers   500000     2   
rivers_83                 - Trans Amadi Port Harcourt Rivers  1000000     2   
rivers_89                  -Trans Amadi Port Harcourt Rivers  1200000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_65       0  0 Toilets  currently being renovated: spacious miniflat (...  
lagos_254      0  1 Toilets  servicedthis is a lovely selfcontain @ akoka y...  
lagos_268      0  0 Toilets  furnisheda newly renovated mini flat @ obadiah...  
lagos_441      0  0 Toilets  room and parlour self contain for rent at dest...  
lagos_442      0  0 Toilets  to let: serviced block of 6-no-3-bedroom flat ...  
...          ...        ...                                                ...  
rivers_71      0  0 Toilets  spacious brand new 1 bedroom for rent in off o...  
rivers_75      0  0 Toilets  beautiful 1 bedroom flat for rent in woji port...  
rivers_82      0  0 Toilets  neat 2 bedroom flat for rent in woji port harc...  
rivers_83      0  0 Toilets  decent king size 2 bedroom flat for rent in pe...  
rivers_89      0  0 Toilets  decent 2 bedroom flat for rent in peter odili ...  

[542 rows x 6 columns]

I'll replace the baths' value with their corresponding toilets' digits, as some of the toilets have a non-zero digit.

In [35]:
data.loc[data.baths == 0, 'baths'] = data.toilets.map(lambda x: x.split(' ')[0] if len(x.split(' '))  > 1 else 0).astype(int)

data[data.baths == 0]

neighborhood    price  beds  \
city_id                                                                       
lagos_65   Close to 3rd Mainland Bridge, Yaba, Lagos. Yab...   800000     1   
lagos_268         Obadia- off ilaje roadBariga Shomolu Lagos   350000     1   
lagos_441     destiny home estate Abijo AJAHAbijo Ajah Lagos   500000     1   
lagos_442  Ajiran Road, Agungi, off Lekki-Epe Express Way...  2000000     3   
lagos_446                                ...Onike Yaba Lagos  1500000     2   
...                                                      ...      ...   ...   
rivers_71               ....Trans Amadi Port Harcourt Rivers  1000000     1   
rivers_75                 - Trans Amadi Port Harcourt Rivers   750000     1   
rivers_82                ...Trans Amadi Port Harcourt Rivers   500000     2   
rivers_83                 - Trans Amadi Port Harcourt Rivers  1000000     2   
rivers_89                  -Trans Amadi Port Harcourt Rivers  1200000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_65       0  0 Toilets  currently being renovated: spacious miniflat (...  
lagos_268      0  0 Toilets  furnisheda newly renovated mini flat @ obadiah...  
lagos_441      0  0 Toilets  room and parlour self contain for rent at dest...  
lagos_442      0  0 Toilets  to let: serviced block of 6-no-3-bedroom flat ...  
lagos_446      0  0 Toilets  executively built brand new 2bedroom apartment...  
...          ...        ...                                                ...  
rivers_71      0  0 Toilets  spacious brand new 1 bedroom for rent in off o...  
rivers_75      0  0 Toilets  beautiful 1 bedroom flat for rent in woji port...  
rivers_82      0  0 Toilets  neat 2 bedroom flat for rent in woji port harc...  
rivers_83      0  0 Toilets  decent king size 2 bedroom flat for rent in pe...  
rivers_89      0  0 Toilets  decent 2 bedroom flat for rent in peter odili ...  

[520 rows x 6 columns]

If the bedroom is just 1, then the bathroom has to be one too.

In [36]:
data.loc[data.baths == 0, 'baths'] = data.beds.apply(lambda x: x if x == 1 else 0)

data[data.baths == 0]

neighborhood    price  beds  \
city_id                                                                       
lagos_442  Ajiran Road, Agungi, off Lekki-Epe Express Way...  2000000     3   
lagos_446                                ...Onike Yaba Lagos  1500000     2   
lagos_451                               -Adekunle Yaba Lagos  1800000     3   
lagos_454                              ...Kosofe Ikosi Lagos   600000     2   
lagos_458                                   ...Gbagada Lagos  1300000     3   
...                                                      ...      ...   ...   
rivers_56  AT VALLEY VIEW ESTATE OFF TANKObia Akpor Port ...   500000     2   
rivers_61                ...Trans Amadi Port Harcourt Rivers   800000     3   
rivers_82                ...Trans Amadi Port Harcourt Rivers   500000     2   
rivers_83                 - Trans Amadi Port Harcourt Rivers  1000000     2   
rivers_89                  -Trans Amadi Port Harcourt Rivers  1200000     2   

           baths    toilets                                              extra  
city_id                                                                         
lagos_442      0  0 Toilets  to let: serviced block of 6-no-3-bedroom flat ...  
lagos_446      0  0 Toilets  executively built brand new 2bedroom apartment...  
lagos_451      0  0 Toilets  brand new 3 bedroom apartment for rent at adek...  
lagos_454      0  0 Toilets  tolet oworosoki letting very decent 2bedroom f...  
lagos_458      0  0 Toilets  *fresh letting* decent 3 bedrooms up all rooms...  
...          ...        ...                                                ...  
rivers_56      0  0 Toilets  this is a very spacious 2 bedroom with spaciou...  
rivers_61      0  0 Toilets  modern king size 3 bedroom flat for rent in wo...  
rivers_82      0  0 Toilets  neat 2 bedroom flat for rent in woji port harc...  
rivers_83      0  0 Toilets  decent king size 2 bedroom flat for rent in pe...  
rivers_89      0  0 Toilets  decent 2 bedroom flat for rent in peter odili ...  

[319 rows x 6 columns]

In [37]:
data[data.baths==0].index.map(lambda x: x[:4]).value_counts()

lago    198
abuj     51
oyo_     42
rive     24
ogun      4
Name: city_id, dtype: int64

Common sense says I should remove the remaining data, but I have Port-Harcourt, where I have just barely 100 rows of data. Losing 24 rows ,out of 100, would be painful so I'd just fill them in with the number of bedrooms (this should be accurate in most cases except in the instances where there's an additional bathroom for the sitting room)

In [38]:
data.loc[data.baths == 0, 'baths'] = data.beds

data.beds.value_counts()

1    2126
3    1637
2    1577
4     267
5     247
Name: beds, dtype: int64

Great! Now I'll delete the `toilets` column

In [39]:
data.drop('toilets', axis=1, inplace=True)

data.sample(5)

neighborhood    price  beds  \
city_id                                                                        
lagos_2433                             Ago Palace Okota Lagos   700000     2   
lagos_4254                          Lekki Phase 1 Lekki Lagos  2500000     4   
abuja_68                            Utako districtUtako Abuja  3000000     3   
lagos_2253  Modile Way Off Akerele RoadRandle Avenue Surul...   700000     1   
abuja_644                                      FO1Kubwa Abuja   800000     2   

            baths                                              extra  
city_id                                                               
lagos_2433      3  furnished newly builtnewly built executive lov...  
lagos_4254      2  servicedfully service 2 bedroom flat for let l...  
abuja_68        3  serviced newly builta tastefully finished, ser...  
lagos_2253      1  newly builtminiflats with pop ceiling with til...  
abuja_644       2  servicedrecently vacated 2bedroom flat, pop fi...

## Feature Engineering

Now I'll extract new features from `city_id`, `neighborhood`, and `extra`

In [40]:
data.neighborhood = data.neighborhood.str.lower() # for consistency

### Extracting State from `city_id`

In [41]:
data['state'] = data.index.map(lambda x: x.split('_')[0].capitalize())

data.head()

neighborhood    price  beds  \
city_id                                                                     
lagos_0         olaleye estate alaka iponri surulere lagos  2500000     3   
lagos_1  off kudirat abiola way, julie estateoregun ike...  2500000     3   
lagos_3                       ...lekki phase 1 lekki lagos   800000     1   
lagos_4                          lekki phase 1 lekki lagos  1200000     1   
lagos_5                                         ketu lagos  1000000     2   

         baths                                              extra  state  
city_id                                                                   
lagos_0      4  brand new service 3 bedroom flat all rooms ens...  Lagos  
lagos_1      3  the property is a 3 bedrooms upscale flat all ...  Lagos  
lagos_3      1  for more information, pricing, and private vie...  Lagos  
lagos_4      1  for more information, pricing, and private vie...  Lagos  
lagos_5      2  brand new 2 bedrooms flat for rent at ketu lag...  Lagos

In [42]:
data.state.value_counts()

Lagos     4390
Abuja      703
Oyo        499
Ogun       183
Rivers      79
Name: state, dtype: int64

In [43]:
data[(data.state == 'Oyo') & (~data.neighborhood.str.contains('ibadan'))]

neighborhood    price  beds  \
city_id                                                                     
oyo_26                                            -ido oyo   200000     1   
oyo_299                                            oyo oyo   150000     1   
oyo_345                                            oyo oyo   160000     3   
oyo_352                                            oyo oyo   220000     2   
oyo_355                                            oyo oyo   350000     3   
oyo_358                                         egbeda oyo   400000     3   
oyo_373                                            ido oyo  2000000     5   
oyo_398                                            ido oyo  2000000     5   
oyo_466   elepe area beside barracks road ojooakinyele oyo   250000     3   
oyo_470  idi ape. beside army barracks fence. akoboolor...   400000     3   

         baths                                              extra state  
city_id                                                                  
oyo_26       1  furnished newly builtnewly built room and parl...   Oyo  
oyo_299      1  furnisheda room and parlor self con at unique ...   Oyo  
oyo_345      3  furnished3 bedroom flat at ore ofe off elewuro...   Oyo  
oyo_352      2  furnished2 bedroom flat at ilupeju estate off ...   Oyo  
oyo_355      3  furnished3 bedroom flat @ ojurin akobo ibadan....   Oyo  
oyo_358      3  furnished newly built3 bedroom flat / apartmen...   Oyo  
oyo_373      4  furnished newly built4 bedroom flat 5 toilet &...   Oyo  
oyo_398      4  furnished newly built4 bedroom flat 5 toilet &...   Oyo  
oyo_466      3             beside barracks road ojoo arulogun ..    Oyo  
oyo_470      3  furnished serviced newly builtits a modern day...   Oyo

10 of the oyo rows are not in Ibadan! Ibadan seems to be in the `description` column of some, though.

In [44]:
non_ibadan_corrected = data[(data.state == 'Oyo') & (~data.neighborhood.str.contains('ibadan')) & (data.extra.str.contains('ibadan'))]
non_ibadan_corrected

neighborhood   price  beds  baths  \
city_id                                     
oyo_26      -ido oyo  200000     1      1   
oyo_299      oyo oyo  150000     1      1   
oyo_345      oyo oyo  160000     3      3   
oyo_352      oyo oyo  220000     2      2   
oyo_355      oyo oyo  350000     3      3   

                                                     extra state  
city_id                                                           
oyo_26   furnished newly builtnewly built room and parl...   Oyo  
oyo_299  furnisheda room and parlor self con at unique ...   Oyo  
oyo_345  furnished3 bedroom flat at ore ofe off elewuro...   Oyo  
oyo_352  furnished2 bedroom flat at ilupeju estate off ...   Oyo  
oyo_355  furnished3 bedroom flat @ ojurin akobo ibadan....   Oyo

Now I'll extract out the address in the `extra` column.

In [45]:
data.loc[non_ibadan_corrected.index, 'neighborhood'] = data.apply(lambda x: x.extra.split(' at ', 1)[1] if ' at ' in x.extra else x.neighborhood, 1)  # where 'at' precedes the address
data.loc[non_ibadan_corrected.index, 'neighborhood'] = data.apply(lambda x: x.extra.split(' @ ', 1)[1] if ' @ ' in x.extra else x.neighborhood, 1)  # where '@' precedes the address

data.loc[non_ibadan_corrected.index]

neighborhood   price  beds  \
city_id                                                                    
oyo_26   apete arola, apete ibadan... tiles all through...  200000     1   
oyo_299  unique estate elewuro ojurin akobo ibadan... r...  150000     1   
oyo_345  ore ofe off elewuro ojurin akobo ibadan... ren...  160000     3   
oyo_352  ilupeju estate off isokan ojurin akobo ibadan....  220000     2   
oyo_355  ojurin akobo ibadan.... rent 300k... tiles, en...  350000     3   

         baths                                              extra state  
city_id                                                                  
oyo_26       1  furnished newly builtnewly built room and parl...   Oyo  
oyo_299      1  furnisheda room and parlor self con at unique ...   Oyo  
oyo_345      3  furnished3 bedroom flat at ore ofe off elewuro...   Oyo  
oyo_352      2  furnished2 bedroom flat at ilupeju estate off ...   Oyo  
oyo_355      3  furnished3 bedroom flat @ ojurin akobo ibadan....   Oyo

Great! Finally, I'll clean off the righthand side so the address would end at Ibadan

In [46]:
data.loc[non_ibadan_corrected.index, 'neighborhood'] = data.neighborhood.map(lambda x: x.split('...')[0])

data.loc[non_ibadan_corrected.index]

neighborhood   price  beds  baths  \
city_id                                                                       
oyo_26                       apete arola, apete ibadan  200000     1      1   
oyo_299      unique estate elewuro ojurin akobo ibadan  150000     1      1   
oyo_345        ore ofe off elewuro ojurin akobo ibadan  160000     3      3   
oyo_352  ilupeju estate off isokan ojurin akobo ibadan  220000     2      2   
oyo_355                            ojurin akobo ibadan  350000     3      3   

                                                     extra state  
city_id                                                           
oyo_26   furnished newly builtnewly built room and parl...   Oyo  
oyo_299  furnisheda room and parlor self con at unique ...   Oyo  
oyo_345  furnished3 bedroom flat at ore ofe off elewuro...   Oyo  
oyo_352  furnished2 bedroom flat at ilupeju estate off ...   Oyo  
oyo_355  furnished3 bedroom flat @ ojurin akobo ibadan....   Oyo

Now I'll delete off the rest that don't have 'Ibadan' in their address

In [47]:
non_ibadan = data[(data.state == 'Oyo') & (~data.neighborhood.str.contains('ibadan'))]
non_ibadan

neighborhood    price  beds  \
city_id                                                                     
oyo_358                                         egbeda oyo   400000     3   
oyo_373                                            ido oyo  2000000     5   
oyo_398                                            ido oyo  2000000     5   
oyo_466   elepe area beside barracks road ojooakinyele oyo   250000     3   
oyo_470  idi ape. beside army barracks fence. akoboolor...   400000     3   

         baths                                              extra state  
city_id                                                                  
oyo_358      3  furnished newly built3 bedroom flat / apartmen...   Oyo  
oyo_373      4  furnished newly built4 bedroom flat 5 toilet &...   Oyo  
oyo_398      4  furnished newly built4 bedroom flat 5 toilet &...   Oyo  
oyo_466      3             beside barracks road ojoo arulogun ..    Oyo  
oyo_470      3  furnished serviced newly builtits a modern day...   Oyo

In [48]:
data.drop(non_ibadan.index, inplace=True)

### Getting ZipCodes, TownNames, & Well-defined Addresses via Google Maps API

In [49]:
import googlemaps 

gmaps = googlemaps.Client(key='AIzaSy********************************')

In [50]:
data.iloc[0]['neighborhood']

'olaleye estate alaka iponri surulere lagos'

In [51]:
returned_map = gmaps.geocode(data.iloc[3]['neighborhood'])
returned_map

[{'address_components': [{'long_name': 'Lekki Phase 1',
    'short_name': 'Lekki Phase 1',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': 'Lekki',
    'short_name': 'Lekki',
    'types': ['locality', 'political']},
   {'long_name': 'Lagos',
    'short_name': 'LA',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'Nigeria',
    'short_name': 'NG',
    'types': ['country', 'political']}],
  'formatted_address': 'Lekki Phase 1, Lekki, Nigeria',
  'geometry': {'bounds': {'northeast': {'lat': 6.4703045, 'lng': 3.4894716},
    'southwest': {'lat': 6.4300926, 'lng': 3.452561}},
   'location': {'lat': 6.447809299999999, 'lng': 3.4723495},
   'location_type': 'APPROXIMATE',
   'viewport': {'northeast': {'lat': 6.4703045, 'lng': 3.4894716},
    'southwest': {'lat': 6.4300926, 'lng': 3.452561}}},
  'place_id': 'ChIJS9Q72lL0OxARKJ3cGrQfM0c',
  'types': ['political', 'sublocality', 'sublocality_level_1']}]

In [52]:
locality, lga, hood, zipcode = np.nan, np.nan, np.nan, np.nan
for info in returned_map[-1]['address_components']:
    locality = info['long_name'] if 'locality' in info['types'] else locality
    lga = info['long_name'] if 'administrative_area_level_2' in info['types'] else lga
    hood = info['long_name'] if ('neighborhood' in info['types'] or 'administrative_area_level_3' in info['types'] or 
                                 'sublocality' in info['types'] or 'colloquial area' in info['types']) else hood
    zipcode = info['long_name'] if 'postal_code' in info['types'] else zipcode
    
print(locality, lga, hood, zipcode)

Lekki nan Lekki Phase 1 nan


No postal code in the giant dictionary.
Now I'll create a function and map it to neighborhood then extract the new features.

In [55]:
def get_gmap_details(row):
    geocode_results = gmaps.geocode(row)
    locality, lga, hood, zipcode = (np.nan, np.nan, np.nan, np.nan)
    try:
        for info in geocode_results[-1]['address_components']:
            locality = info['long_name'] if 'locality' in info['types'] else locality
            lga = info['long_name'] if 'administrative_area_level_2' in info['types'] else lga
            hood = info['long_name'] if ('neighborhood' in info['types'] or 'administrative_area_level_3' in info['types'] 
                                         or 'sublocality' in info['types'] or 'colloquial area' in info['types']) else hood
            zipcode = info['long_name'] if 'postal_code' in info['types'] else zipcode
        print(f"{locality}, {lga}, {hood}, {zipcode}")
    except IndexError:
        print('No details gotten!')
    return (locality, lga, hood, zipcode)

In [72]:
data[['locality', 'lga', 'hood', 'zipcode']] = data.neighborhood.map(get_gmap_details).to_list()

Lagos, Surulere, Oke Ira, nan
Ikeja, Ikeja, Allen, nan
Lekki, nan, Lekki Phase 1, nan
Lekki, nan, Lekki Phase 1, nan
Lagos, Kosofe, Ketu, nan
Lekki, nan, nan, nan
nan, Eti Osa, Ajah/Sangotedo, nan
Ojodu Berger, nan, nan, nan
Lagos, Surulere, Orile Iganmu, nan
Lagos, Shomolu, Abule ijesha, 100001
Lagos, Ikeja, Ojodu, nan
Lagos, Lagos Mainland, Onike, nan
Lagos, Shomolu, Igbobi, 100001
Lagos, Shomolu, Igbobi, 100246
Lagos, Kosofe, Gbagada, nan
Ikeja, Ikeja, Ojodu, 100213
Lagos, Ifako/Ijaye, nan, nan
nan, Ibeju Lekki, nan, nan
Ojodu, nan, Ogba, nan
Lekki, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Eti-Osa, nan
Lagos, Surulere, nan, nan
Lekki, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Eti-Osa, nan
Lagos, Eti Osa, Lekki Phase I, 234001
Lekki, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Eti-Osa, nan
Sangotedo, Eti Osa, Eti-Osa, nan
Lekki, Eti Osa, Lekki Penninsula II, nan
nan, Ibeju Lekki, nan, nan
Lekki, Ibeju Lekki, nan, nan
Lekki, Ibeju Lekki, na

In [73]:
data.head()

neighborhood    price  beds  \
city_id                                                                     
lagos_0         olaleye estate alaka iponri surulere lagos  2500000     3   
lagos_1  off kudirat abiola way, julie estateoregun ike...  2500000     3   
lagos_3                       ...lekki phase 1 lekki lagos   800000     1   
lagos_4                          lekki phase 1 lekki lagos  1200000     1   
lagos_5                                         ketu lagos  1000000     2   

         baths                                              extra  state  \
city_id                                                                    
lagos_0      4  brand new service 3 bedroom flat all rooms ens...  Lagos   
lagos_1      3  the property is a 3 bedrooms upscale flat all ...  Lagos   
lagos_3      1  for more information, pricing, and private vie...  Lagos   
lagos_4      1  for more information, pricing, and private vie...  Lagos   
lagos_5      2  brand new 2 bedrooms flat for rent at ketu lag...  Lagos   

        locality       lga           hood zipcode  
city_id                                            
lagos_0    Lagos  Surulere        Oke Ira     NaN  
lagos_1    Ikeja     Ikeja          Allen     NaN  
lagos_3    Lekki       NaN  Lekki Phase 1     NaN  
lagos_4    Lekki       NaN  Lekki Phase 1     NaN  
lagos_5    Lagos    Kosofe           Ketu     NaN

I'll save the data as a CSV file so as not to go through the stress of calling the Google Maps API again

In [74]:
data.to_csv('data_with_gmap_features.csv', index=False)